src/main/java/net/oauth/jsontoken/AsyncJsonTokenParser.java

```diff
@@ -125,7 +153,7 @@ public AsyncJsonTokenParser(
       }
       futureVerifiers = provider.findVerifier(jsonToken.getIssuer(), jsonToken.getKeyId());
     } catch (Exception e) {
-      return Futures.immediateFailedFuture(e);
+      return getFailedFuture(e);
```

Comment:
Instead of plugging in getFailedFuture throughout each step, is it possible to add a final step to the public methods using Futures.catching or Futures.catchingAsync to do the exception mapping?

https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Futures.html#catching(com.google.common.util.concurrent.ListenableFuture,%20java.lang.Class,%20com.google.common.base.Function,%20java.util.concurrent.Executor)

Reply:
With how the class is right now, we can add it. However, if we map exceptions from the existing public methods, I don't think we would need to do any or complicated exception rethrowing in the new async methods.

Reply:
Since I changed the method signatures of the three functions that the async functions rely on, there is now minimal rethrowing. All of the rethrowing occurs in synchronous functions (eg. verify, deserialize, signatureIsValid) or the first part of provideVerifiers (before any asynchronous functions). Should we still use Futures.catching-, given the simplified rethrowing?

Reply:
My worry is that a simple change in the future could mess this up, by introducing some code that needs the exception to be mapped, but forgetting to add the call to map exceptions. Adding a call like "return mapExceptions(originalFuture)" at the end of the public methods is a simple way to prevent that. It also makes it more consistent since all the public methods will have the mapExceptions at the end, instead of only the synchronous ones that don't return futures.

Reply:
Changed.